File: backend/app/custom_fields.py
"""Opportunity custom fields — the definitions, and what an answer may be.

`CustomFieldDef` says a field exists. `Opportunity.custom_fields` holds the answers.
This module is the only place that knows how those two fit together, so the API, the
create path and the detail PATCH cannot drift on what an answer is allowed to be.

Three rules do all the safety work here, and every one of them exists because the
values in that blob are things a customer told somebody on the phone:

1. **An answer is never destroyed by an edit it was not part of.** `merge_answers`
   MERGES: a key the client did not send keeps the value it had. Clearing an answer
   is an explicit act — you send the key with an empty value — not something that
   happens because a form rendered a shorter list of fields than the last one did.

2. **Hidden is not gone.** A field attached to pipelines A and B is not shown on a
   deal in pipeline C, and neither is an archived field. Their answers stay in the
   blob untouched and reappear the moment the deal moves back or the field is
   un-archived. They also cannot be *written* from a form that is not showing them,
   which is refused out loud rather than ignored.

3. **The `owen_` namespace belongs to the telephony project.** No definition may
   claim a key in it, and no write through this module can drop one of those keys.
   `owen_call_id` keeps its existing guard: it is the join key, and changing it
   breaks call attribution in a separate production system (DECISIONS.md).
"""
import math
from datetime import date

from fastapi import HTTPException
from sqlalchemy import select
from sqlalchemy.orm import Session, selectinload

from .models import CustomFieldDef, CustomFieldPipeline

# Namespaces this app does not own. A key under one of these is written by a
# machine, read by a machine, and shown to a person — never edited by one.
#
#   owen_*    the telephony project ("from OWEN"). See DECISIONS.md — the four keys
#             were measured on the live GHL account before any of this was built.
#   workiz_*  the Workiz migration (`app/workiz_import.py`, 2026-09-11).
#             `workiz_id` is the Client # on a contact and the Job # on an
#             opportunity, and it is the whole of the import's idempotency: the
#             importer finds a record by it and UPDATES rather than duplicating.
#             Editing one by hand would silently make the next import create a
#             second copy of that customer, so the API refuses. A record created in
#             the CRM by hand simply has no workiz_id, which is expected.
#
# One tuple, checked in one function, so a third namespace is a one-line change and
# not a second guard that drifts from this one.
RESERVED_PREFIXES = ("owen_", "workiz_")
# Kept as a name because it reads better in the message `claim_key` raises, and
# because it is the prefix that has been in this codebase the longest.
RESERVED_PREFIX = RESERVED_PREFIXES[0]
# The one key that is a live join key into another production system's data.
JOIN_KEY = "owen_call_id"
# The import's identity key. Unlike `owen_call_id`, which may be SET where there was
# none, this one is immutable in every direction through this module — see
# `_check_reserved`.
IMPORT_PREFIX = "workiz_"

KEY_MAX = 64
LABEL_MAX = 160
TEXT_MAX = 2000
OPTION_MAX = 80
OPTIONS_MAX = 50

TEXT = CustomFieldDef.TEXT
NUMBER = CustomFieldDef.NUMBER
DROPDOWN = CustomFieldDef.DROPDOWN
DATE = CustomFieldDef.DATE
BOOLEAN = CustomFieldDef.BOOLEAN
TYPES = CustomFieldDef.TYPES

# Accepted spellings for a yes/no answer. The browser sends "true"/"false"; the CLI
# and a human typing JSON send all of these, and refusing "yes" for a field labelled
# "yes/no" would be a joke at the user's expense.
_TRUE = {"true", "yes", "y", "1", "on"}
_FALSE = {"false", "no", "n", "0", "off"}


def is_reserved(key: str) -> bool:
    return key.startswith(RESERVED_PREFIXES)


def slug_for(label: str) -> str:
    """The storage key for a label, derived ONCE at creation and then immutable.

    The key is what every recorded answer is filed under, so a later rename of the
    label must not touch it — otherwise editing "How old is the roof?" into "Roof
    age" would orphan every answer already given. ASCII only: the key ends up in a
    JSON object, a URL and a CSV header, and a smart-quote in it helps nobody.
    """
    flat = "".join(c if (c.isascii() and c.isalnum()) else "_" for c in label.lower())
    return "_".join(part for part in flat.split("_") if part)[:KEY_MAX].strip("_")


# ---------- reading definitions ----------

def load_defs(db: Session, *, include_archived: bool = True) -> list[CustomFieldDef]:
    stmt = select(CustomFieldDef).options(selectinload(CustomFieldDef.pipelines))
    if not include_archived:
        stmt = stmt.where(CustomFieldDef.archived_at.is_(None))
    # `id` is the tiebreak for the same reason the board uses one: `position`
    # defaults to 0, so two fields can share it and the order would otherwise
    # change between two identical reads.
    return list(db.scalars(
        stmt.order_by(CustomFieldDef.position, CustomFieldDef.id)).all())


def pipeline_ids(d: CustomFieldDef) -> list[int]:
    return sorted(link.pipeline_id for link in d.pipelines)


def shows_on(d: CustomFieldDef, pipeline_id: int | None) -> bool:
    """Is this field asked on a deal in `pipeline_id`?

    Attachment is explicit. A field attached to no pipeline shows on no deal —
    an empty set never means "everywhere", so detaching the last pipeline cannot
    silently turn "AHS claim number" into a question on every job.
    """
    if d.archived_at is not None or pipeline_id is None:
        return False
    return any(link.pipeline_id == pipeline_id for link in d.pipelines)


def describe(d: CustomFieldDef) -> dict:
    return {
        "id": d.id,
        "key": d.key,
        "label": d.label,
        "field_type": d.field_type,
        "options": list(d.options or []),
        "position": d.position,
        "entity": d.entity,
        "archived": d.archived_at is not None,
        "archived_at": d.archived_at,
        "pipeline_ids": pipeline_ids(d),
    }


# ---------- defining a field ----------

def clean_label(label: str | None) -> str:
    cleaned = (label or "").strip()
    if not cleaned:
        raise HTTPException(400, "a custom field needs a question to ask")
    if len(cleaned) > LABEL_MAX:
        raise HTTPException(400, "a custom field's question is limited to %d characters"
                            % LABEL_MAX)
    return cleaned


def clean_type(field_type: str | None) -> str:
    if field_type not in TYPES:
        raise HTTPException(400, "%r is not a field type — expected one of %s" % (
            field_type, ", ".join(TYPES)))
    return field_type


def clean_options(field_type: str, options: list | None) -> list[str]:
    """Only a dropdown has options, and a dropdown without them offers nothing."""
    values = [str(o).strip() for o in (options or []) if str(o).strip()]
    if field_type != DROPDOWN:
        return []
    if not values:
        raise HTTPException(400, "a dropdown needs at least one option to choose from")
    if len(values) > OPTIONS_MAX:
        raise HTTPException(400, "a dropdown is limited to %d options" % OPTIONS_MAX)
    for v in values:
        if len(v) > OPTION_MAX:
            raise HTTPException(400, "the option %r is longer than %d characters"
                                % (v, OPTION_MAX))
    seen, unique = set(), []
    for v in values:
        if v.lower() in seen:
            continue
        seen.add(v.lower())
        unique.append(v)
    return unique


def claim_key(db: Session, label: str) -> str:
    """Derive the storage key, and refuse the two ways it can be wrong.

    The reserved refusal is the safety requirement. A field defined as
    `owen_campaign` would put a form control on top of live call-attribution data;
    a field defined as `workiz_id` would put one on top of the key the importer
    matches records by, and the next import would then create a second copy of
    every customer somebody had edited. Note the check is on the DERIVED key, so
    labelling a field "Owen campaign" or "Workiz ID" is refused too — that is the
    case a prefix check on the label alone would have let through.
    """
    key = slug_for(label)
    if not key:
        raise HTTPException(400, "that question has no letters or digits in it, so "
                                 "there is nothing to store the answers under")
    if is_reserved(key):
        owner = ("the telephony project — those fields are written by OWEN"
                 if key.startswith(RESERVED_PREFIX)
                 else "the Workiz migration — those fields are written by the "
                      "importer")
        raise HTTPException(400, (
            "%r starts with %r, which belongs to %s and cannot be redefined here. "
            "Try a different wording." % (
                key, next(p for p in RESERVED_PREFIXES if key.startswith(p)),
                owner)))
    clash = db.scalar(select(CustomFieldDef).where(CustomFieldDef.key == key))
    if clash:
        raise HTTPException(409, (
            "%r is already the key for %r%s. Two fields cannot share a key — the "
            "answers would land on top of each other." % (
                key, clash.label, " (archived)" if clash.archived_at else "")))
    return key


# ---------- answers ----------

def _shown(value) -> str:
    text = value if isinstance(value, str) else repr(value)
    return text if len(text) <= 40 else text[:37] + "..."


def _reject(d: CustomFieldDef, value, expected: str) -> None:
    raise HTTPException(400, "“%s” takes %s, and “%s” is not one." % (
        d.label, expected, _shown(value)))


def coerce_answer(d: CustomFieldDef, raw):
    """The stored form of one answer, or None for "not answered".

    Validation is by TYPE, at the API, so the CLI and the telephony feed get the
    same answer as the browser. An empty string is not a failed number — it is a
    question nobody has answered yet, and it is stored by removing the key.
    """
    if raw is None or (isinstance(raw, str) and not raw.strip()):
        return None

    if d.field_type == NUMBER:
        # `bool` is an `int` in Python, so True would otherwise store as 1.
        if isinstance(raw, (bool, list, dict)):
            _reject(d, raw, "a number")
        if isinstance(raw, (int, float)):
            number = float(raw)
        else:
            try:
                number = float(str(raw).strip())
            except ValueError:
                _reject(d, raw, "a number")
        if not math.isfinite(number):
            _reject(d, raw, "a number")
        # 3.0 came in as "3"; store it as 3 so the CSV and the form agree.
        return int(number) if number.is_integer() else number

    if d.field_type == DROPDOWN:
        value = str(raw).strip()
        options = list(d.options or [])
        if value not in options:
            raise HTTPException(400, "“%s” does not offer “%s”. The choices are: %s."
                                % (d.label, _shown(value), ", ".join(options) or "none"))
        return value

    if d.field_type == DATE:
        value = str(raw).strip()
        try:
            date.fromisoformat(value)
        except ValueError:
            _reject(d, raw, "a date like 2026-09-11")
        return value

    if d.field_type == BOOLEAN:
        if isinstance(raw, bool):
            return raw
        word = str(raw).strip().lower()
        if word in _TRUE:
            return True
        if word in _FALSE:
            return False
        _reject(d, raw, "yes or no")

    # TEXT
    if isinstance(raw, (list, dict)):
        _reject(d, raw, "text")
    value = str(raw).strip()
    if len(value) > TEXT_MAX:
        raise HTTPException(400, "“%s” is limited to %d characters." % (d.label, TEXT_MAX))
    return value


def merge_answers(db: Session, *, pipeline_id: int | None,
                  existing: dict | None, incoming: dict | None) -> dict:
    """The blob an opportunity should hold after this write.

    MERGE, not replace. What that buys, in order of how badly it would hurt:

    * A reserved `owen_*` key can never be dropped, whatever the client sends.
      Deleting `owen_campaign` would silently break call attribution, and until now
      any client that posted a shorter object did exactly that.
    * An answer to a field that is not shown on THIS deal — wrong pipeline, or
      archived — is kept exactly as it stands. That is what makes "move it to
      another pipeline and back and the answer is still there" true.
    * An answer the client simply did not mention is left alone, rather than read
      as a deletion.

    Clearing is explicit: send the key with "" or null.
    """
    existing = dict(existing or {})
    incoming = dict(incoming or {})
    defs = {d.key: d for d in load_defs(db)}
    result: dict = {}

    # --- the reserved namespace ---
    was = existing.get(JOIN_KEY)
    now = incoming.get(JOIN_KEY, was)
    if was is not None and now != was:
        raise HTTPException(
            400, "owen_call_id is the telephony join key and cannot be changed here")
    # `workiz_*` is stricter than `owen_call_id`, which may still be SET where there
    # was none. These are read-only in every direction: not set, not changed, not
    # removed. The importer writes them through the ORM and nothing else may. An
    # unchanged echo is not a write — the detail form posts the whole blob back on
    # every save — so only a real difference is refused, and it is refused out loud
    # rather than ignored, because silently discarding a value somebody typed is how
    # an edit looks like it worked when it did not.
    for key in sorted(set(existing) | set(incoming)):
        if not key.startswith(IMPORT_PREFIX):
            continue
        if key in incoming and incoming[key] != existing.get(key):
            raise HTTPException(400, (
                "%s is written by the Workiz import and is read-only here — it is "
                "how a re-import recognises this record instead of creating a "
                "second copy of it." % key))
    # Kept first, and unconditionally: no path through this function drops one.
    for key, value in existing.items():
        if is_reserved(key):
            result[key] = value
    for key, value in incoming.items():
        if is_reserved(key):
            result[key] = value

    # --- defined fields ---
    for key, d in defs.items():
        if shows_on(d, pipeline_id):
            if key not in incoming:
                if key in existing:
                    result[key] = existing[key]
            elif incoming[key] == existing.get(key):
                # An unchanged echo is not a write and is never re-validated. It
                # matters: an admin who removes a dropdown option would otherwise
                # make every deal already holding that answer unsaveable, because
                # the form posts the whole object back on every save.
                result[key] = existing[key]
            else:
                answer = coerce_answer(d, incoming[key])
                if answer is not None:
                    result[key] = answer
            continue
        # Not shown here. The answer survives untouched...
        if key in existing:
            result[key] = existing[key]
        # ...and cannot be written from a form that is not showing the field.
        # Refused out loud: silently ignoring a value somebody typed is how an
        # answer goes missing without anyone noticing. An echo of what is already
        # stored is not a write, and neither is a blank where nothing is stored —
        # a client that posts the whole blob back must not be refused for it.
        if key in incoming and incoming[key] != existing.get(key) and not (
                key not in existing and (incoming[key] is None or (
                    isinstance(incoming[key], str) and not incoming[key].strip()))):
            raise HTTPException(400, (
                "“%s” is %s, so its answer cannot be set here. The answer already "
                "recorded is kept." % (
                    d.label,
                    "archived" if d.archived_at is not None
                    else "not asked on this pipeline")))

    # --- everything else ---
    # The column was free-form JSON before any of this existed and stays that way:
    # a key no definition claims is stored verbatim. `null` removes one, which is
    # the only way an undefined key can be taken out.
    for key, value in incoming.items():
        if key in defs or is_reserved(key):
            continue
        if value is None:
            continue
        result[key] = value
    for key, value in existing.items():
        if key in defs or is_reserved(key) or key in incoming:
            continue
        result[key] = value

    return result


def attach(db: Session, d: CustomFieldDef, ids: list[int] | None) -> None:
    """Replace the field's pipeline attachments. Caller validates the ids exist."""
    wanted = set(ids or [])
    for link in list(d.pipelines):
        if link.pipeline_id not in wanted:
            db.delete(link)
        else:
            wanted.discard(link.pipeline_id)
    for pipeline_id in sorted(wanted):
        db.add(CustomFieldPipeline(field_id=d.id, pipeline_id=pipeline_id))
    db.flush()

File: frontend/src/lib/customFields.ts
/**
 * Which job questions a deal is asked, and how one answer is drawn.
 *
 * Import-free on purpose, like calendarGrid.ts and reminders.ts, so
 * backend/tests/test_custom_fields_ui.py can run it under node and assert real
 * behaviour instead of grepping the source. That matters more here than usual:
 * the whole feature turns on a field being HIDDEN without its answer being lost,
 * and "hidden" is a decision this file makes.
 *
 * The server is still the thing that enforces all of it — `merge_answers` in
 * backend/app/custom_fields.py keeps an answer whatever a client sends. Nothing
 * here is a gate; it decides what to draw.
 */

export type FieldType = 'text' | 'number' | 'dropdown' | 'date' | 'boolean' | 'paragraph'

/** What a yes/no question may show beside its tick (2026-09-14). */
export type LinkedField = 'contact_email' | 'opportunity_address'

export type FieldDef = {
  id: number
  key: string
  label: string
  field_type: FieldType
  options: string[]
  position: number
  entity: string
  archived: boolean
  pipeline_ids: number[]
  /** The modal tab it is drawn under; null = Opportunity details. Optional only
      so fixtures written before groups existed still type-check. */
  group_id?: number | null
  /** The call Checklist's settings (2026-09-14). Optional so older fixtures
      type-check; the server always sends all three. */
  script?: string | null
  /** yes/no only: the REAL value drawn and edited beside the tick. */
  linked_field?: LinkedField | null
  /** dropdown only: the choices that open a details box. */
  details_when?: string[]
}

/** The details box's answer is stored beside the main one. Mirrors DETAILS_SUFFIX
    in backend/app/custom_fields.py; no derived key can contain "__". */
export const DETAILS_SUFFIX = '__details'

export function detailsKey(key: string): string {
  return key + DETAILS_SUFFIX
}

/** Is the details box open for this answer? Only when the CHOSEN option is one the
    field says calls for details — never for a blank, never for another option. */
export function showsDetails(def: FieldDef, value: unknown): boolean {
  if (def.field_type !== 'dropdown' || !def.details_when?.length) return false
  return typeof value === 'string' && def.details_when.includes(value)
}

/** The tab whose progress the board card shows. Mirrors CHECKLIST_GROUP. */
export const CHECKLIST_GROUP = 'Checklist'

export function isChecklistGroup(name: string): boolean {
  return name.trim().toLowerCase() === CHECKLIST_GROUP.toLowerCase()
}

/**
 * "N / M answered" for one tab. M is the questions this pipeline asks in it (the
 * caller passes exactly those — `modalSections` already dropped the rest), N the
 * ones holding an answer. `false` is an answer; a details box is not a question.
 * The server's `custom_fields.progress` counts the same way for the card badge.
 */
export function answeredCount(fields: FieldDef[], answers: Record<string, unknown>):
  { answered: number; total: number } {
  return {
    answered: fields.filter((d) => !isEmptyAnswer(answers[d.key])).length,
    total: fields.length,
  }
}

/** A question drawn across both columns: long text, a script, a linked value or a
    details box would be cramped into half the modal. */
export function spansRow(def: FieldDef): boolean {
  return def.field_type === 'paragraph' || !!def.script?.trim() || !!def.linked_field
    || !!def.details_when?.length
}

/**
 * Namespaces this app does not own: `owen_*` (the telephony project), `workiz_*`
 * (the Workiz import) and `ahs_job_*` (the AHS email relay, 2026-09-14). The values
 * stay on every deal untouched — `owen_call_id` is a live join key, `workiz_id` the
 * import's idempotency key and `ahs_job_id` the relay's — but the owner does not
 * want them on the screen (2026-09-13), so nothing in the modal renders one — except
 * `workiz_tech`, read-only, by the owner's later decision (see `workizTechnicians`).
 * Mirrors `RESERVED_PREFIXES` in backend/app/custom_fields.py.
 */
export const RESERVED_PREFIXES = ['owen_', 'workiz_', 'ahs_job_'] as const
export const JOIN_KEY = 'owen_call_id'

export function isReserved(key: string): boolean {
  return RESERVED_PREFIXES.some((p) => key.startsWith(p))
}

/**
 * The questions asked on a deal in this pipeline, in display order.
 *
 * Attachment is explicit: a field attached to no pipeline is asked nowhere. An
 * empty list never means "everywhere", so detaching the last pipeline cannot turn
 * a narrow question into a global one by accident. A reserved key is never asked,
 * even if a definition somehow claimed one — the server refuses to create it.
 */
export function askedOn(defs: FieldDef[], pipelineId: number | null | undefined):
  FieldDef[] {
  if (pipelineId == null) return []
  return defs
    .filter((d) => !d.archived && d.pipeline_ids.includes(pipelineId))
    .filter((d) => !isReserved(d.key))
    .sort((a, b) => a.position - b.position || a.id - b.id)
}

export type GroupLike = { id: number; name: string; position: number }

/**
 * The modal's left nav, as data: the fields with no group (drawn under
 * Opportunity details) and one entry per group, IN GROUP ORDER, each holding the
 * fields of that group this pipeline asks.
 *
 * Every group appears whatever the pipeline, because a group applies to all of
 * them; one whose fields are all on other pipelines is an empty tab, not a
 * missing one. A field pointing at a group that no longer exists falls back to
 * Opportunity details rather than vanishing — the same place the server puts it.
 */
export function modalSections(
  defs: FieldDef[], groups: GroupLike[], pipelineId: number | null | undefined,
): { ungrouped: FieldDef[]; groups: { group: GroupLike; fields: FieldDef[] }[] } {
  const asked = askedOn(defs, pipelineId)
  const known = new Set(groups.map((g) => g.id))
  const ordered = [...groups].sort((a, b) => a.position - b.position || a.id - b.id)
  return {
    ungrouped: asked.filter((d) => d.group_id == null || !known.has(d.group_id)),
    groups: ordered.map((group) => ({
      group, fields: asked.filter((d) => d.group_id === group.id),
    })),
  }
}

/** Is an answer blank, for "Hide empty fields"? `false` is an answer; `''` is not. */
export function isEmptyAnswer(value: unknown): boolean {
  return value === undefined || value === null
    || (typeof value === 'string' && value.trim() === '')
}

/**
 * The answers an Update should send: only the ones that CHANGED, and never a
 * reserved key. The modal holds the whole blob so its controls can read it, but
 * posting that blob back would put `owen_*` and `workiz_*` values on the wire on
 * every save for no reason. The server keeps them either way; this keeps the
 * request honest about what the user did. A blank where nothing was stored is not
 * a change.
 */
export function changedAnswers(
  stored: Record<string, unknown>, current: Record<string, unknown>,
): Record<string, unknown> {
  const out: Record<string, unknown> = {}
  for (const [key, value] of Object.entries(current)) {
    if (isReserved(key)) continue
    if (value === stored[key]) continue
    if (isEmptyAnswer(value) && isEmptyAnswer(stored[key])) continue
    out[key] = value
  }
  return out
}

/** What goes in the input for one field. Booleans become the select's values. */
export function inputValue(def: FieldDef, value: unknown): string {
  if (value === undefined || value === null) return ''
  if (def.field_type === 'boolean') {
    return value === true ? 'true' : value === false ? 'false' : ''
  }
  return String(value)
}

/**
 * What the input hands back, ready to post.
 *
 * An empty control is `''` and NOT null: the server reads both as "unanswered",
 * and `''` keeps the object shape the form rendered, which is what makes
 * clearing an answer an explicit act rather than an omission.
 */
export function submitValue(def: FieldDef, raw: string): string | number | boolean {
  if (raw === '') return ''
  if (def.field_type === 'boolean') return raw === 'true'
  if (def.field_type === 'number') {
    // Left as the typed string when it is not a number, so the SERVER refuses it
    // with its own sentence naming the field. Silently dropping it here would
    // save the deal and lose what the user typed without saying anything.
    const n = Number(raw)
    return raw.trim() !== '' && Number.isFinite(n) ? n : raw
  }
  return raw
}

/** A stored answer as a person reads it. Used for the read-only rows. */
export function displayValue(def: FieldDef | null, value: unknown): string {
  if (value === null || value === undefined || value === '') return '--'
  if (def?.field_type === 'boolean' || typeof value === 'boolean') {
    return value ? 'Yes' : 'No'
  }
  return String(value)
}

/**
 * A dropdown's options, with the stored answer added if it is no longer offered.
 *
 * Precedent: the appointment panel keeps an unlisted status as an option so
 * opening a booking cannot silently rewrite it. Same reasoning — retiring "Tile"
 * must not turn every deal holding it into a deal holding "Shingle" the first
 * time somebody opens it.
 */
export function optionsFor(def: FieldDef, value: unknown): string[] {
  const stored = value === null || value === undefined ? '' : String(value)
  if (stored && !def.options.includes(stored)) return [...def.options, stored]
  return def.options
}

/** "Inspection — Tue 9:00am", the way a deal names the visit booked for it. */
export function describeBooking(
  title: string, startsAt: string, locale = 'en-US',
): string {
  const at = new Date(startsAt)
  if (Number.isNaN(at.getTime())) return title
  const when = at.toLocaleString(locale, {
    weekday: 'short', hour: 'numeric', minute: '2-digit',
  })
  return `${title} — ${when}`
}

/**
 * The answers a CREATE should send (the Add new opportunity modal, 2026-09-14): only
 * the questions the chosen pipeline asks, their details boxes, and nothing blank.
 *
 * The dispatcher can answer a question and then switch the Pipeline to one that does
 * not ask it. The server refuses an answer to a question the pipeline does not ask —
 * rightly — so posting the whole form would turn a pipeline change into a refused
 * Create with the caller still on the phone. What is dropped here was never saved.
 */
export function answersFor(
  defs: FieldDef[], pipelineId: number | null | undefined, answers: Record<string, unknown>,
): Record<string, unknown> {
  const out: Record<string, unknown> = {}
  for (const def of askedOn(defs, pipelineId)) {
    for (const key of [def.key, ...(def.field_type === 'dropdown' ? [detailsKey(def.key)] : [])]) {
      if (!isEmptyAnswer(answers[key])) out[key] = answers[key]
    }
  }
  return out
}

/**
 * The one reserved key the modal DOES draw (the owner, 2026-09-15): `workiz_tech`, the
 * Workiz job's technician names, written by the import as a list in the job's order.
 * Shown read-only as "Workiz technician(s)"; '' when the card has none, so "Hide empty
 * fields" hides it. Nothing sends it back — `changedAnswers` skips every reserved key,
 * and the server refuses a change to it.
 */
export const WORKIZ_TECH_KEY = 'workiz_tech'

export function workizTechnicians(customFields: Record<string, unknown> | null | undefined): string {
  const raw = customFields?.[WORKIZ_TECH_KEY]
  const names = Array.isArray(raw) ? raw : typeof raw === 'string' ? [raw] : []
  return names
    .filter((n): n is string => typeof n === 'string')
    .map((n) => n.split(/\s+/).filter(Boolean).join(' '))
    .filter(Boolean)
    .join(', ')
}
